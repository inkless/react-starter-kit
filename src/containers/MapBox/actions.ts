import { ViewState } from 'react-map-gl';
import { ActionType, createAsyncAction, createStandardAction } from 'typesafe-actions';
import { MapboxState, SimulationGeoFeature } from './types';

/**
 * Three actions for sending request to fetch geocoding
 * request, success, failure
 */
export const fetchOriginGeocodingAction = createAsyncAction(
  '@@MapBox/FETCH_ORIGIN_GEOCODING_REQUEST',
  '@@MapBox/FETCH_ORIGIN_GEOCODING_SUCCESS',
  '@@MapBox/FETCH_ORIGIN_GEOCODING_FAILURE'
)<undefined, MapboxState['origin'], Error>();

/**
 * Action to set map viewport
 */
export const setViewport = createStandardAction('@@MapBox/SET_VIEWPORT')<ViewState>();

/**
 * Action to set deliverSimulationStarted, indicating if simulation has started
 */
export const setDeliverSimulationStarted = createStandardAction(
  '@@MapBox/SET_DELIVER_SIMULATION_STARTED'
)<boolean>();

/**
 * Action to set current deliver simulation's geo features,
 * including routes and points (showing the small car)
 */
export const setDeliverGeoFeatures = createStandardAction('@@MapBox/SET_DELIVER_GEO_FEATURES')<
  SimulationGeoFeature[]
>();

/**
 * Move camera to a courier
 */
export const moveCamera = createStandardAction('@@MapBox/MOVE_CAMERA')<string>();

export type MapboxAction =
  | ActionType<typeof fetchOriginGeocodingAction>
  | ActionType<typeof setViewport>
  | ActionType<typeof setDeliverSimulationStarted>
  | ActionType<typeof setDeliverGeoFeatures>
  | ActionType<typeof moveCamera>;
