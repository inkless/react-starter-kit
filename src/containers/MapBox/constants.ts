export const MAPBOX_API_URL = 'https://api.mapbox.com';
export const DEFAULT_ORIGIN = {
  name: 'Home',
  address: '1800 Marine Street, Santa Monica, CA 90405',
};

/**
 * source for mapbox, all our layers is sharing one source,
 * so it has better performance
 */
export const MAPBOX_SOURCE_NAME = 'react-starter-kit';

/**
 * Simulation animation interval
 * Technically using requestAnimationFrame can make animation smoothier,
 * but that consumes too much CPU/GPU, using an interval instead
 */
export const ANIMATION_INTERVAL = 500;
