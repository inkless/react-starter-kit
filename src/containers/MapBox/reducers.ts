import { getType } from 'typesafe-actions';

import { MapboxState } from './types';
import {
  MapboxAction,
  fetchOriginGeocodingAction,
  setViewport,
  setDeliverSimulationStarted,
  setDeliverGeoFeatures,
  moveCamera,
} from './actions';
import { Feature, Point } from '@turf/helpers';

const defaultState: MapboxState = {
  origin: {
    address: '',
    coordinates: [0, 0],
  },
  viewport: {
    longitude: 0,
    latitude: 0,
    zoom: 11.5,
  },
  deliverSimulationStarted: false,
  deliverSimulationGeoFeatures: [],
};

export function mapboxReducer(state = defaultState, action: MapboxAction): MapboxState {
  switch (action.type) {
    case getType(fetchOriginGeocodingAction.success): {
      const [longitude, latitude] = action.payload.coordinates;
      return {
        ...state,
        origin: action.payload,
        viewport: {
          ...state.viewport,
          longitude,
          latitude,
        },
      };
    }
    case getType(setViewport): {
      // Note: this is needed, ReactMapGL will send width and height with the
      // viewport, in order to allow map resize properly when window resize,
      // we need to remove this width, height from payload
      const { width, height, ...viewport } = action.payload;
      return {
        ...state,
        viewport,
      };
    }
    case getType(setDeliverSimulationStarted):
      return {
        ...state,
        deliverSimulationStarted: action.payload,
      };
    case getType(setDeliverGeoFeatures):
      return {
        ...state,
        deliverSimulationGeoFeatures: action.payload,
      };
    case getType(moveCamera): {
      const courierPoint = state.deliverSimulationGeoFeatures.find(
        feature => feature.properties.id === action.payload
      ) as Feature<Point, any> | undefined;

      if (!courierPoint || !courierPoint.geometry) {
        return state;
      }
      const [longitude, latitude] = courierPoint.geometry.coordinates;
      return {
        ...state,
        viewport: {
          ...state.viewport,
          longitude,
          latitude,
        },
      };
    }
    default:
      return state;
  }
}
