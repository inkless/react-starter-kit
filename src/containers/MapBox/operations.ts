import lineSliceAlong from '@turf/line-slice-along';
import { point } from '@turf/helpers';

import { ThunkResult } from '@src/store';
import { DeliveringOrder, PendingOrder } from '@src/containers/OrderList/types';
import {
  startDeliveringOrder,
  fetchDestinationAction,
  orderDelivered,
} from '@src/containers/OrderList/actions';
import {
  fetchOriginGeocodingAction,
  setDeliverSimulationStarted,
  setDeliverGeoFeatures,
} from './actions';
import { DEFAULT_ORIGIN, ANIMATION_INTERVAL } from './constants';
import {
  fetchGeocoding,
  fetchDrivingRoute,
  generateRandomSpeed,
  createRouteFromGeometry,
  calculateRouteLength,
} from './utils';
import { SimulationGeoFeature } from '@src/containers/MapBox/types';

/**
 * @param address A query string of address
 */
export function fetchOriginOperation(address = DEFAULT_ORIGIN.address): ThunkResult<void> {
  return async function(dispatch) {
    const { request, success, failure } = fetchOriginGeocodingAction;
    dispatch(request());
    try {
      const coordinates = await fetchGeocoding(address);
      dispatch(
        success({
          address,
          coordinates,
        })
      );
    } catch (e) {
      dispatch(failure(e));
    }
  };
}

/**
 * Send API request to fetch pendingOrder's geo infomation,
 * then send an action to start delivering order.
 * If simulation is not started, start simulation.
 */
export function fetchDestinationOperation({
  dispatchTime,
  ...order
}: PendingOrder): ThunkResult<void> {
  return async function(dispatch, getState) {
    const {
      mapbox: { origin, deliverSimulationStarted },
    } = getState();
    dispatch(fetchDestinationAction.request());
    try {
      // fetch coordinates first, then fetch driving route info
      const coordinates = await fetchGeocoding(order.destination);
      const { geometry } = await fetchDrivingRoute(origin.coordinates, coordinates);
      const route = createRouteFromGeometry(geometry, {
        id: `route-${order.id}`,
        route: 'destination',
      });

      const deliveringOrder: DeliveringOrder = {
        ...order,
        coordinates,
        route,
        startTime: new Date(),
        speed: generateRandomSpeed(),
        distance: calculateRouteLength(route),
      };

      // start simulation if not started
      if (!deliverSimulationStarted) {
        dispatch(startSimulatingDeliver());
      }

      dispatch(startDeliveringOrder(deliveringOrder));
    } catch (e) {
      dispatch(fetchDestinationAction.failure(e));
    }
  };
}

/**
 * startSimulatingDeliver
 */
export function startSimulatingDeliver(): ThunkResult<void> {
  return function(dispatch, getState) {
    // do nothing if it's already started
    if (getState().mapbox.deliverSimulationStarted) {
      return;
    }

    dispatch(setDeliverSimulationStarted(true));

    /**
     * Notice: To have a even smoothier animation,
     * we could use requestAnimationFrame.
     * However, that consume much more CPU/GPU, I'll keep it simple here.
     */
    const interval = setInterval(() => {
      const {
        orderList: { delivering, pending },
      } = getState();

      const features: SimulationGeoFeature[] = [];
      delivering.forEach(order => {
        const { id, name, speed, startTime, route, distance } = order;

        const traveledDistance = speed * ((Date.now() - startTime.valueOf()) / 1000);

        // Use traveledDistance, calculate current route
        const currentRoute = lineSliceAlong(route, 0, traveledDistance, { units: 'miles' });
        currentRoute.properties = {
          ...currentRoute.properties,
          id: `current-route-${id}`,
          route: 'current',
        };

        if (currentRoute.geometry !== null) {
          // the position of courier should be the last point of currentRoute
          const lastCoordinate = currentRoute.geometry.coordinates.slice(-1)[0];
          const currentPoint = point(lastCoordinate, {
            name,
            id: `current-point-${id}`,
          });

          // push point/route to features
          features.push(currentPoint, currentRoute);
        }

        // dispatch orderDelivered when travel distance is greater than
        // destination distance
        if (traveledDistance >= distance) {
          dispatch(
            orderDelivered({
              ...order,
              endTime: new Date(),
            })
          );
        }
      });

      // dispatch geo features
      if (features.length) {
        dispatch(setDeliverGeoFeatures(features));
      }

      // when there's no more pending, and delivering,
      // stop simulation
      if (!pending.length && !delivering.length) {
        clearInterval(interval);
      }
    }, ANIMATION_INTERVAL);
  };
}
