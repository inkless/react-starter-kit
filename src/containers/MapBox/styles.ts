import BASE_MAP_STYLE from '@src/data/mapbox-style-streets-v11.json';
import { MAPBOX_SOURCE_NAME } from './constants';
import { colors } from '@src/theme';

/**
 * A map style inherit from streets-v11, while adding our own layers
 */
export const mapStyle = {
  ...BASE_MAP_STYLE,
  layers: [
    ...BASE_MAP_STYLE.layers,
    {
      // layer for destination route, colored in gray,
      // indicating where the courier should go
      id: 'ck-route-layer',
      source: MAPBOX_SOURCE_NAME,
      type: 'line',
      paint: {
        'line-width': 2,
        'line-color': colors.grays[5],
        'line-opacity': 0.5,
      },
      filter: ['all', ['in', '$type', 'LineString'], ['in', 'route', 'destination']],
    },
    {
      // layer for courier's current route, colored in blue
      id: 'ck-current-route',
      source: MAPBOX_SOURCE_NAME,
      type: 'line',
      paint: {
        'line-color': colors.common.blue,
        'line-width': 2,
      },
      filter: ['all', ['in', '$type', 'LineString'], ['in', 'route', 'current']],
    },
    {
      // layer for courier itself, show as a car, with text-field showing the
      // name of the food
      id: 'ck-current-point',
      source: MAPBOX_SOURCE_NAME,
      type: 'symbol',
      layout: {
        'icon-image': 'car-15',
        'icon-rotate': ['get', 'bearing'],
        'icon-rotation-alignment': 'map',
        'icon-allow-overlap': true,
        'icon-ignore-placement': true,
        'text-field': ['get', 'name'],
        'text-size': 12,
        'text-offset': [0, -1.5],
        'text-allow-overlap': true,
      },
      paint: {
        'text-color': '#202',
        'text-halo-color': '#fff',
        'text-halo-width': 2,
      },
      filter: ['all', ['in', '$type', 'Point']],
    },
  ],
};
