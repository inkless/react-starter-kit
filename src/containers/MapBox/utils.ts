import request from '@src/utils/request';
import { lineString, Feature } from '@turf/helpers';
import turfLength from '@turf/length';
import { decode } from '@mapbox/polyline';

import { StringAnyMap } from '@src/types/utils';
import { MAPBOX_API_URL } from './constants';

const ACCESS_TOKEN = process.env.MapboxAccessToken;
const POLYLINE_PRECISION = 5;
const SECONDS_IN_HOUR = 3600;
const SPEED_BASE = 20;

/**
 * Fetch driving route from origin to destination
 *
 * @param origin
 * @param destination
 */
export async function fetchDrivingRoute(origin: [number, number], destination: [number, number]) {
  const baseUrl = `${MAPBOX_API_URL}/directions/v5/mapbox/driving-traffic`;
  const coordinates = `${origin.join(',')};${destination.join(',')}`;
  const url = `${baseUrl}/${coordinates}?steps=true&overview=full&access_token=${ACCESS_TOKEN}`;
  const data = await request(url);
  return data.routes[0];
}

/**
 * Given any address, find the geocoding
 * @param address
 */
export async function fetchGeocoding(address: string) {
  const url = `${MAPBOX_API_URL}/geocoding/v5/mapbox.places/${encodeURIComponent(
    address
  )}.json?access_token=${ACCESS_TOKEN}`;
  const data = await request(url);
  return data.features[0].center;
}

/**
 * Generate a speed between 20 to 25 miles per hour
 * @returns miles per second
 */
export function generateRandomSpeed() {
  return (Math.random() * 5 + SPEED_BASE) / SECONDS_IN_HOUR;
}

/**
 * Decode geometry and return geojson route
 * @param geometry
 * @returns geojson lineString
 */
export function createRouteFromGeometry(geometry: string, options?: StringAnyMap) {
  const decoded = decode(geometry, POLYLINE_PRECISION).map(c => {
    return c.reverse();
  });

  return lineString(decoded, options);
}

/**
 * Given a route, calculate the distance of it
 * @param route
 * @returns distance in miles
 */
export function calculateRouteLength(route: Feature) {
  return turfLength(route, { units: 'miles' });
}
