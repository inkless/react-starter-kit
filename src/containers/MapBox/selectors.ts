import { createSelector } from 'reselect';
import { featureCollection } from '@turf/helpers';

import { RootState } from '@src/store';
import { MAPBOX_SOURCE_NAME } from './constants';
import { mapStyle } from './styles';

export const getDeliveringOrders = (state: RootState) => state.orderList.delivering;

export const getDeliveringGeoFeatures = (state: RootState) =>
  state.mapbox.deliverSimulationGeoFeatures;

/**
 * Based on deliveringOrders and deliveringGeoFeatures, create a source which
 * can be consumed by MapboxGL
 * We don't want this data to be in redux because otherwise it
 * will be duplicated data with deliverSimulationGeoFeatures.
 */
export const sourceSelector = createSelector(
  getDeliveringOrders,
  getDeliveringGeoFeatures,
  (orders, features) => ({
    type: 'geojson',
    data: featureCollection([...orders.map(order => order.route), ...features]),
  })
);

/**
 * select mapStyle, we only update sources in the style
 * mapStyle is never stored in redux because:
 * 1. it's too big
 * 2. the majority part of the data never change
 * So we only update part of the mapStyle when needed.
 */
export const mapStyleSelector = createSelector(
  sourceSelector,
  source => ({
    ...mapStyle,
    sources: {
      ...mapStyle.sources,
      [MAPBOX_SOURCE_NAME]: source,
    },
  })
);
