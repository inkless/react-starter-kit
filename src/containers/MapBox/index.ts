import { connect } from 'react-redux';

import MapBox from '@src/components/MapBox';
import { RootState } from '@src/store';
import { fetchOriginOperation } from './operations';
import { setViewport } from './actions';
import { mapStyleSelector, getDeliveringOrders } from './selectors';

const mapStateToProps = (state: RootState) => ({
  ...state.mapbox,
  mapStyle: mapStyleSelector(state),
  deliveringOrders: getDeliveringOrders(state),
});

const mapDispatchToProps = {
  fetchOriginOperation,
  setViewport,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(MapBox);
