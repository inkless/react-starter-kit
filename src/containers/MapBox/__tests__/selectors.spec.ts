import { RootState } from '@src/store';
import { createInitialState } from '@src/utils/test-helper';
import { sourceSelector, mapStyleSelector } from '@src/containers/MapBox/selectors';
import { delivering } from '@src/components/OrderList/__tests__/__fixture__/orders';
import { fixtureData1 } from '@src/containers/MapBox/__tests__/__fixture__/geojson';

const rootState = createInitialState();

describe('MapBox Selectors', () => {
  let state: RootState;
  beforeEach(() => {
    state = {
      ...rootState,
      orderList: {
        ...rootState.orderList,
        delivering,
      },
      mapbox: {
        ...rootState.mapbox,
        deliverSimulationGeoFeatures: [fixtureData1.route],
      },
    };
  });

  it('should select source correctly', () => {
    const source = sourceSelector(state);
    expect(source).toMatchSnapshot();
  });

  it('should select mapStyle correctly', () => {
    const mapStyle = mapStyleSelector(state);
    expect(mapStyle.sources).toMatchSnapshot();
  });
});
