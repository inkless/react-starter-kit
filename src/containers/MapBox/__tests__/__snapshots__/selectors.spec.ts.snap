// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapBox Selectors should select mapStyle correctly 1`] = `
Object {
  "composite": Object {
    "type": "vector",
    "url": "mapbox://mapbox.mapbox-streets-v8,mapbox.mapbox-terrain-v2",
  },
  "react-starter-kit": Object {
    "data": Object {
      "features": Array [
        Object {
          "geometry": Object {
            "coordinates": Array [
              Array [
                -118.46742,
                34.00534,
              ],
              Array [
                -118.47105,
                34.00321,
              ],
            ],
            "type": "LineString",
          },
          "properties": Object {},
          "type": "Feature",
        },
        Object {
          "geometry": Object {
            "coordinates": Array [
              Array [
                -118.46742,
                34.00534,
              ],
              Array [
                -118.47105,
                34.00321,
              ],
            ],
            "type": "LineString",
          },
          "properties": Object {},
          "type": "Feature",
        },
        Object {
          "geometry": Object {
            "coordinates": Array [
              Array [
                -118.46742,
                34.00534,
              ],
              Array [
                -118.47105,
                34.00321,
              ],
            ],
            "type": "LineString",
          },
          "properties": Object {},
          "type": "Feature",
        },
        Object {
          "geometry": Object {
            "coordinates": Array [
              Array [
                -118.46742,
                34.00534,
              ],
              Array [
                -118.47105,
                34.00321,
              ],
            ],
            "type": "LineString",
          },
          "properties": Object {},
          "type": "Feature",
        },
        Object {
          "geometry": Object {
            "coordinates": Array [
              Array [
                -118.46742,
                34.00534,
              ],
              Array [
                -118.47105,
                34.00321,
              ],
            ],
            "type": "LineString",
          },
          "properties": Object {},
          "type": "Feature",
        },
        Object {
          "geometry": Object {
            "coordinates": Array [
              Array [
                -118.46742,
                34.00534,
              ],
              Array [
                -118.47105,
                34.00321,
              ],
            ],
            "type": "LineString",
          },
          "properties": Object {},
          "type": "Feature",
        },
        Object {
          "geometry": Object {
            "coordinates": Array [
              Array [
                -118.46742,
                34.00534,
              ],
              Array [
                -118.47105,
                34.00321,
              ],
            ],
            "type": "LineString",
          },
          "properties": Object {},
          "type": "Feature",
        },
        Object {
          "geometry": Object {
            "coordinates": Array [
              Array [
                -118.46742,
                34.00534,
              ],
              Array [
                -118.47105,
                34.00321,
              ],
            ],
            "type": "LineString",
          },
          "properties": Object {},
          "type": "Feature",
        },
        Object {
          "geometry": Object {
            "coordinates": Array [
              Array [
                -118.46742,
                34.00534,
              ],
              Array [
                -118.47105,
                34.00321,
              ],
            ],
            "type": "LineString",
          },
          "properties": Object {},
          "type": "Feature",
        },
        Object {
          "geometry": Object {
            "coordinates": Array [
              Array [
                -118.46742,
                34.00534,
              ],
              Array [
                -118.47105,
                34.00321,
              ],
            ],
            "type": "LineString",
          },
          "properties": Object {},
          "type": "Feature",
        },
        Object {
          "geometry": Object {
            "coordinates": Array [
              Array [
                -118.46742,
                34.00534,
              ],
              Array [
                -118.47105,
                34.00321,
              ],
              Array [
                -118.47115,
                34.00309,
              ],
              Array [
                -118.47325,
                34.00463,
              ],
              Array [
                -118.47629,
                34.00289,
              ],
              Array [
                -118.48058,
                33.99993,
              ],
              Array [
                -118.48113,
                33.99954,
              ],
              Array [
                -118.48088,
                33.9993,
              ],
              Array [
                -118.48088,
                33.99928,
              ],
            ],
            "type": "LineString",
          },
          "properties": Object {},
          "type": "Feature",
        },
      ],
      "type": "FeatureCollection",
    },
    "type": "geojson",
  },
}
`;

exports[`MapBox Selectors should select source correctly 1`] = `
Object {
  "data": Object {
    "features": Array [
      Object {
        "geometry": Object {
          "coordinates": Array [
            Array [
              -118.46742,
              34.00534,
            ],
            Array [
              -118.47105,
              34.00321,
            ],
          ],
          "type": "LineString",
        },
        "properties": Object {},
        "type": "Feature",
      },
      Object {
        "geometry": Object {
          "coordinates": Array [
            Array [
              -118.46742,
              34.00534,
            ],
            Array [
              -118.47105,
              34.00321,
            ],
          ],
          "type": "LineString",
        },
        "properties": Object {},
        "type": "Feature",
      },
      Object {
        "geometry": Object {
          "coordinates": Array [
            Array [
              -118.46742,
              34.00534,
            ],
            Array [
              -118.47105,
              34.00321,
            ],
          ],
          "type": "LineString",
        },
        "properties": Object {},
        "type": "Feature",
      },
      Object {
        "geometry": Object {
          "coordinates": Array [
            Array [
              -118.46742,
              34.00534,
            ],
            Array [
              -118.47105,
              34.00321,
            ],
          ],
          "type": "LineString",
        },
        "properties": Object {},
        "type": "Feature",
      },
      Object {
        "geometry": Object {
          "coordinates": Array [
            Array [
              -118.46742,
              34.00534,
            ],
            Array [
              -118.47105,
              34.00321,
            ],
          ],
          "type": "LineString",
        },
        "properties": Object {},
        "type": "Feature",
      },
      Object {
        "geometry": Object {
          "coordinates": Array [
            Array [
              -118.46742,
              34.00534,
            ],
            Array [
              -118.47105,
              34.00321,
            ],
          ],
          "type": "LineString",
        },
        "properties": Object {},
        "type": "Feature",
      },
      Object {
        "geometry": Object {
          "coordinates": Array [
            Array [
              -118.46742,
              34.00534,
            ],
            Array [
              -118.47105,
              34.00321,
            ],
          ],
          "type": "LineString",
        },
        "properties": Object {},
        "type": "Feature",
      },
      Object {
        "geometry": Object {
          "coordinates": Array [
            Array [
              -118.46742,
              34.00534,
            ],
            Array [
              -118.47105,
              34.00321,
            ],
          ],
          "type": "LineString",
        },
        "properties": Object {},
        "type": "Feature",
      },
      Object {
        "geometry": Object {
          "coordinates": Array [
            Array [
              -118.46742,
              34.00534,
            ],
            Array [
              -118.47105,
              34.00321,
            ],
          ],
          "type": "LineString",
        },
        "properties": Object {},
        "type": "Feature",
      },
      Object {
        "geometry": Object {
          "coordinates": Array [
            Array [
              -118.46742,
              34.00534,
            ],
            Array [
              -118.47105,
              34.00321,
            ],
          ],
          "type": "LineString",
        },
        "properties": Object {},
        "type": "Feature",
      },
      Object {
        "geometry": Object {
          "coordinates": Array [
            Array [
              -118.46742,
              34.00534,
            ],
            Array [
              -118.47105,
              34.00321,
            ],
            Array [
              -118.47115,
              34.00309,
            ],
            Array [
              -118.47325,
              34.00463,
            ],
            Array [
              -118.47629,
              34.00289,
            ],
            Array [
              -118.48058,
              33.99993,
            ],
            Array [
              -118.48113,
              33.99954,
            ],
            Array [
              -118.48088,
              33.9993,
            ],
            Array [
              -118.48088,
              33.99928,
            ],
          ],
          "type": "LineString",
        },
        "properties": Object {},
        "type": "Feature",
      },
    ],
    "type": "FeatureCollection",
  },
  "type": "geojson",
}
`;
